### PR TITLE
Proxy onUnload request in full-screen interactive [#181169438]

### PR DESCRIPTION
**Pull request by dougmartin**
This change adds a proxy of the getInteractiveState unloading request from Lara/AP to the wrapped interactive.

### CODE REVIEW

**Review by kswenson — COMMENTED** [review]
It seems likely that this sort of proxying will be required in other contexts. It's worth thinking about whether there's a way to encapsulate any of the proxying behavior in a reusable custom hook.

File: src/shared/components/iframe-runtime.tsx [hunk 1]
```diff
@@ -45,14 +45,31 @@ export const IframeRuntime: React.FC<IProps> =
...
+
   useEffect(() => {
     const initInteractive = () => {
       const phone = phoneRef.current;
       if (!phone) {
         return;
       }
+
+      // Handle proxying the unloading request from Lara/AP through to the wrapped interactive.
+      // The promise that is saved is resolved in the subsequent interactiveState listener callback
```
[the remark this answers is not in the document]
Comment (dougmartin) [reply src/shared/components/iframe-runtime.tsx:61]: Done.

File: src/shared/components/iframe-runtime.tsx [hunk 2]
```diff
@@ -45,14 +45,31 @@ export const IframeRuntime: React.FC<IProps> =
...
+      setOnUnload((options: IGetInteractiveState) => {
+        if (options.unloading) {
+          return new Promise<any>(resolve => {
+            resolveOnUnload.current = resolve;
+            // send the unload request to the wrapped interactive
+            phone.post("getInteractiveState", options);
```
[the remark this answers is not in the document]
Comment (dougmartin) [reply src/shared/components/iframe-runtime.tsx:67]: Yeah, it was being dropped on the floor.  The full-screen interactive was just listening and maintaining the interactive state of the wrapped interactive.

**Comment by dougmartin** [comment]
@kswenson this will all need to be re-evaulated to come up with a generic solution for all the interactives that wrap other interactives.

**Comment by scytacki** [comment]
I think in the case of the full screen interactive we could do something more low level and just proxy almost every request and response. The one set of apis it might have to mess with are the sizing apis. But everything else can probably just pass through. This way we don't need to add new code to the full screen interactive for every new or changed api.

But for the other consumers of this iframe-runtime (side-by-side, carousel, scaffolded question), they all need to support multiple interactives and in some cases they also save their own state. I think in these cases we probably need to add some support at the host level probably both for runtime and authoring hosts. 

An example is have better support in the reports. For the carousel and side-by-side the report should really show each nested item as a separate item in the report. That means each of these nested interactives should be treated like a top level interactive in the report. So that means each nested item gets its own answer document in firestore. If we support that properly then we could have the host send messages specific to each nested interactive through the top-level interactive. And each nested interactive can send messages through the top-level interactive back to the host. So it is basically the proxying stuff again, but now the host is aware of the proxying.  But this will be complicated because the top-level interactive will need to probably mess with the size, and the new unload feature, and in the case of scaffolded question it probably needs to do more customization. But I'm hopeful that we can set it up so the top-level interactive doesn't need to know about every possible message just the ones it needs to customize.

**Review by scytacki — APPROVED** [review]
I think this looks good as is.

**Pull request merged**